### DISCUSSION TITLE
Response is always nullable under `generateNullableReferenceTypes`?

### DISCUSSION THREAD

**Discussion by jods4** [christianhelle/refitter#211] [Q&A] [answered]
Given this basic OpenApi:
```json
"get": {
  "responses": {
    "200": {
      "content": {
        "application/json": {
          "schema": {
            "$ref": "#/components/schemas/response"
          }
```

And when option `generateNullableReferenceTypes` is true, then the following method is generated:

```cs
[Get("/")]
Task<Response?> Get();
```

Why is `Response` nullable? Is my OpenApi missing a property to indicate the response is always present?
When the API is successful (200) there _is_ a `Response` object.

With this code generation I basically have to always bang `!` my Refit responses.

**Comment by christianhelle** [comment]
@jods4 thanks for bringing this up. The option `generateNullableReferenceTypes` is an NSwag specific option, and I just 100% piggy-back over NSwag for parsing the OpenAPI document. It seems that setting `generateNullableReferenceTypes` always generates null reference types. But in the case of the generated `Refit` interface, I tend to use the `Task<IApiResponse<T>>` return type to avoid exceptions for failed responses

All that said, since Refitter generates the Refit interface, if a `get` method will never return `null` and instead throw an exception, then we can ignore the fact the NSwag says it returns a nullable type, and just declare the Refit interface method to return a non-nullable type.

I'll play around with this a bit and see if we can find a good solution

**Reply by jods4** [reply]
Hi @christianhelle,
Did you find something to make the return type in interface non-nullable on your side?
Do you think I should open an issue on NSwag instead (I'd like to be sure that it's 100% on NSwag side, as I'm not using it directly)?

**Reply by christianhelle** [reply]
@jods4 I unfortunately completely forgot about this issue. 

NSwag just released a recent major version update that might include what you need. I'll try to revisit this tonight after I put the family to bed

**Reply by jods4** [reply]
@christianhelle I tried with latest Refitter -- which I believe uses NSwag 14? -- and nothing changed.

**Comment by christianhelle** [comment]
@jods4 I looked into this a bit last night and can see that NSwag itself generates its interface as

```csharp
        /// <summary>Find pet by ID</summary>
        /// <param name="petId">ID of pet to return</param>
        /// <returns>successful operation</returns>
        /// <exception cref="ApiException">A server side error occurred.</exception>
        System.Threading.Tasks.Task<Pet> GetPetByIdAsync(long petId);
```

So I looked into the code a bit and found

![image](https://github.com/christianhelle/refitter/assets/710400/5cacc691-422c-4f1f-a01b-a31274d59fcc)

When I set the `isNullable` argument to `false` then the Refit interface is generated as

```csharp
        /// <summary>Find pet by ID</summary>
        /// <remarks>Returns a single pet</remarks>
        /// <param name="petId">ID of pet to return</param>
        /// <returns>successful operation</returns>
        /// <throws cref="ApiException">Thrown when the request returns a non-success status code.</throws>
        [Headers("Accept: application/json")]
        [Get("/pet/{petId}")]
        Task<Pet> GetPetById(long petId);
```

even though I set `generateNullableReferenceTypes` to `true` in the refitter settings

This looks like a bug in Refitter that is trivial to fix. My little concern here is that Refitter users have most likely just worked around this issue and introducing this "fix" would "break" things for others

**Reply by jods4** [reply, marked as answer]
I understand your concerns; we had to "fix" this ourselves as well.
I don't see many ways to work around the current codegen: we either use `?.` or `!` when accessing the response (I guess you could also check `x == null`).

So I just verified this on [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgYRIG8CT6zT1sAGEmCMAUxIF4SARAIDcdBmPrEyrEgCEuAMwD2AJy4AKCgH4SsAM4AXCDADGXAJQ0JDelAUl1+o6Z69+MAK4AbL5fQB2QRFrGwDdGENjMy0AOg5uEmRwyJcAQjjOLlF8GwBfa2spZjYAQQUDLhVNZOczS1ocmwY7Byco13dvXzJAoWymhjC2l1j4niThs3Sx/oZ8/FygA): none of those patterns warn when the type is changed to a non-nullable reference.

I guess the risk of breaking things is low?

Long term I think this is something worth fixing because when working with nullable reference the current behavior is just annoying and objectively wrong. If you want zero risk you could provide a setting to configure the codegen but that's maintenance burden for you.

**Reply by christianhelle** [reply]
@jods4 I agree that this should be fixed, it will cause a minor inconvenience to those who have implemented a work-around. You are right that in the long term its better to fix it

I created issue #302 and I will fix this today

**Reply by jods4** [reply]
Awesome, thanks a lot!

**Reply by christianhelle** [reply]
@jods4 You're welcome. Thanks for bringing this up again. The traction and positive that I received for Refitter can be a bit overwhelming and it's a bit hard to follow every single thing. I completely forgot to follow up on this one so I'm sorry for the delay and inconvenience

I will release the fix once [this pull request](https://github.com/christianhelle/refitter/pull/304) gets merged in

**Reply by jods4** [reply]
No worries! I actually had forgotten about this as well but stumbled upon it again when writing new code.

Refit is a great way to manually write API clients, and Refitter is a super helpful companion tool when OpenAPI definitions are available. Thanks! ❤

**Comment by christianhelle** [comment]
The fix for this is released in [version 0.9.6](https://github.com/christianhelle/refitter/releases/tag/0.9.6)
